### PR TITLE
Daden/sumdist pr Add Distributed Sampler for Iterable Dataset and Example for Distributed Training on AML for Extractive Summarization 

### PR DESCRIPTION
**Pull request by daden-ms**
### Description
Add Distributed Sampler for Iterable Dataset
Example Notebook for Distributed Training on AML for Extractive Summarization 

### Checklist:

- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](https://github.com/microsoft/nlp-recipes/blob/master/CONTRIBUTING.md).
- [x] I have added tests.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: utils_nlp/models/transformers/extractive_summarization.py [hunk 1]
```diff
@@ -96,12 +172,12 @@ def __init__(self, file_list, is_shuffle=False):
                 files is shuffled when the dataset is loaded. Defaults to False.
         """
 
-        self.file_list = file_list
+        self.file_list = sorted(file_list)
```
Comment (saidbleik) [inline utils_nlp/models/transformers/extractive_summarization.py]: Is sorting really needed here? Might conflict with users' expectation in case the list is provided in a certain order.

File: utils_nlp/models/transformers/extractive_summarization.py [hunk 2]
```diff
@@ -16,7 +18,7 @@
 from transformers import BertModel, DistilBertModel
 
 from bertsum.models import data_loader, model_builder
-from bertsum.models.data_loader import Batch
+from bertsum.models.data_loader import Batch,DataIterator
```
Comment (saidbleik) [inline utils_nlp/models/transformers/extractive_summarization.py]: black

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb [hunk 3]
```diff
@@ -0,0 +1,795 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create the workspace using the specified parameters\n",
+    "ws = Workspace.create(name = WORKSPACE_NAME,\n",
```
Comment (saidbleik) [inline examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: azureml_utils.py has a get_or_create_workspace function if you expect the workspace to exist sometimes...

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb [hunk 4]
```diff
@@ -0,0 +1,795 @@
...
+      "setup() is now deprecated. Instead, please use create() to create a new workspace, or get()/from_config() to retrieve an existing one\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Workspace name: daden1amlws\n",
+      "Workspace region: eastus2\n",
+      "Subscription id: 9086b59a-02d7-4687-b3fd-e39fa5e0fd9b\n",
```
Comment (saidbleik) [inline examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: you might want to clear sensitive info (subscription ID, etc...)

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb [hunk 5]
```diff
@@ -0,0 +1,795 @@
...
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#AZUREML_CONFIG_PATH = \"./.azureml\"\n",
+    "## output dir in the workspace\n",
+    "MODEL_NAME = \"distilbert-base-uncased\"\n",
+    "ENCODER = \"transformer\"\n",
+    "TARGET_OUTPUT_DIR = f'output/{EXPERIMENT_NAME}/'\n",
```
Comment (saidbleik) [inline examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: Can we group all imports and constants and place them at the top of the notebook

File: utils_nlp/models/transformers/extractive_summarization.py [hunk 6]
```diff
@@ -26,6 +28,75 @@
...
+
+    def __iter__(self):
+        return self.sampler.iter(self.eachiter())
+
+
+    def _next_dataset_iterator(self, dataset_iter):
+        try:
+            # Drop the current dataset for decreasing memory
+            if hasattr(self, "cur_dataset"):
+                self.cur_dataset = None
```
Comment (hlums) [inline utils_nlp/models/transformers/extractive_summarization.py]: Curious why `del self.cur_dataset` and `gc.collect()` is not enough. I guess there must be a reason.

File: utils_nlp/models/transformers/common.py [hunk 7]
```diff
@@ -212,16 +220,20 @@ def predict(self, eval_dataloader, get_inputs, num_gpus, gpu_ids, verbose=True):
...
 
-    def save_model(self):
-        output_model_dir = os.path.join(self.cache_dir, "fine_tuned")
-
-        os.makedirs(self.cache_dir, exist_ok=True)
-        os.makedirs(output_model_dir, exist_ok=True)
-
-        logger.info("Saving model checkpoint to %s", output_model_dir)
+    def save_model(self, full_name=None):
         # Save a trained model, configuration and tokenizer using `save_pretrained()`.
```
Comment (hlums) [inline utils_nlp/models/transformers/common.py]: Docstring needs to be updated. If full_name is provided, model is state_dict is saved and can be loaded by torch.load()

File: examples/text_summarization/extractive_summarization_cnndm_distributed_train.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (hlums) [inline examples/text_summarization/extractive_summarization_cnndm_distributed_train.py]: Did you eventually figure out why one of you nodes is always ahead of the other one?

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb [hunk 9]
```diff
@@ -0,0 +1,795 @@
+{
```
Comment (hlums) [inline examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: <p>Please remove this sensitive information from the notebook. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp-recipes/pull/548/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (hlums) [inline examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: <p>Same here. Remove the sensitive information from the output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/nlp-recipes/pull/548/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

File: examples/text_summarization/extractive_summarization_cnndm_distributed_train.py (hunk 8 above)
Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_distributed_train.py]: The gradient sync is asynchronous

File: utils_nlp/models/transformers/extractive_summarization.py (hunk 6 above)
Comment (daden-ms) [reply utils_nlp/models/transformers/extractive_summarization.py]: Not sure whether it's enough or not. The code is directly copied from the original source.

File: utils_nlp/models/transformers/extractive_summarization.py (hunk 1 above)
Comment (daden-ms) [reply utils_nlp/models/transformers/extractive_summarization.py]: yes, it's required. I have observed if it's not sorted, the evaluation result through AML is very bad.

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb (hunk 4 above)
Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: will do

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb (hunk 5 above)
Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: I had all the constants together but decided to not group them together. In this way, I am clear of what constants and import are needed in what step. Is there a style requirement for the repo. Personally I prefer this way. But if we see strong need of grouping them together, I can change the notebook.

File: utils_nlp/models/transformers/common.py (hunk 7 above)
Comment (daden-ms) [reply utils_nlp/models/transformers/common.py]: will do

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb (hunk 9 above)
Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: will do

Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: will do

File: utils_nlp/models/transformers/extractive_summarization.py [hunk 10]
```diff
@@ -34,21 +105,26 @@ def __init__(self, adict):
         self.__dict__.update(adict)
 
 
-def get_dataloader(data_iter, shuffle=True, is_labeled=False, batch_size=3000):
+def get_dataloader(data_iter, shuffle=True, is_labeled=False, batch_size=3000, world_size=1, rank=1):
```
[the remark this answers is not in the document]
Comment (daden-ms) [reply utils_nlp/models/transformers/extractive_summarization.py]: Yes, the rank should default to -1 in both places.

File: utils_nlp/models/transformers/extractive_summarization.py (hunk 6 above)
Comment (daden-ms) [reply utils_nlp/models/transformers/extractive_summarization.py]: my guess is that line 87 makes sure the object that self.cur_dataset is de-referenced. and the "del" is used to decrease the reference counter. Not sure why "del" needs to be called twice.

File: examples/text_summarization/extractive_summarization_cnndm_distributed_train.py (hunk 8 above)
Comment (hlums) [reply examples/text_summarization/extractive_summarization_cnndm_distributed_train.py]: Is there a reference documentation?

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb (hunk 5 above)
Comment (hlums) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: I think there are two advantages of grouping the constants
- It's clear from one cell what all the constant values are, so it's less likely for someone to forget to change some constants or assume the wrong constant values.
- It will be easier to control all the constants using one one parameter cell in the testing code.

File: examples/text_summarization/extractive_summarization_cnndm_distributed_train.py (hunk 8 above)
Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_distributed_train.py]: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#skewed-processing-speeds

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb (hunk 5 above)
Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: The disadvantage is that it's not clear of what constant is being used at what step as they are all together.  Which testing code are you referring to?

Comment (hlums) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: When we do integration tests, sometimes we control some parameters in the tests to reduce the test running time. For example, https://github.com/microsoft/nlp-recipes/blob/master/tests/integration/test_notebooks_entailment.py. This is done by setting the cell with all the constants a "parameter" cell and pass the parameter values from the test.

Comment (saidbleik) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: I personally prefer grouping them for the reasons @hlums mentioned. You can add a comment before every group of params (for ex *model params*, *dataset params*, etc...) to make it clearer. Anyway, we don't have strong requirements and feel free to keep it that way.

File: utils_nlp/models/transformers/extractive_summarization.py [hunk 11]
```diff
@@ -26,6 +28,75 @@
...
+        rank (int): Rank of the current GPU. Defaults to 1.
+        
+    """
+    def __init__(self, world_size=1, rank=-1):
+        self.world_size = world_size
+        self.rank = rank
+
+    def iter(self, iterable):
+        if self.rank != -1:
+            return itertools.islice(iterable, self.rank, None, self.world_size)
```
[the remark this answers is not in the document]
Comment (daden-ms) [reply utils_nlp/models/transformers/extractive_summarization.py]: islice() does not take keyword arguments

File: examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb (hunk 5 above)
Comment (daden-ms) [reply examples/text_summarization/extractive_summarization_cnndm_aml_distributed.ipynb]: grouped contants together and added comments before each group
